lib/molinillo/resolution.rb

```diff
@@ -213,36 +220,211 @@ def initial_state
+              requirement_state.possibilities.any? { |set| set.dependencies.none? { |dep| dep == r } }
+            if candidate_index > unwind_details[:index] || unwind_details[:relationship] == :primary
+              unwind_details[:index] = candidate_index
+              unwind_details[:relationship] = :parent
+              unwind_details[:requirement] = r
+              unwind_details[:conflicting_requirements] = binding_requirements
+            end
+
+            next # No need to look at this requirement's grandparent, as it couldn't have a higher index
+          end
+
+          # Finally, look at the grandparent and up of this requirement, looking
+          # for any possibilities at all - we're now too far removed from the conflict
+          # to know how any such possibility will affect it, so just have to try it
+          grandparent_r = parent_of(parent_r)
+          until grandparent_r.nil?
+            requirement_state = find_state_for(grandparent_r)
+            candidate_index = states.index(requirement_state)
+            break unless candidate_index && candidate_index >= unwind_details[:index]
+            if requirement_state && !requirement_state.possibilities.empty?
+              unwind_details[:index] = candidate_index
+              unwind_details[:relationship] = :grandparent
+              unwind_details[:requirement] = r
+              unwind_details[:conflicting_requirements] = binding_requirements
               break
             end
-            r = parent_of(r)
+            grandparent_r = parent_of(grandparent_r)
           end
         end
 
-        index
+        unwind_details
+      end
+
+      # @param [DependencyState] state
+      # @param [Array] array of requirements
+      # @return [Boolean] whether or not the given state has any possibilities
+      #    that could satisfy the given requirements
+      def conflict_fixing_possibilities?(state, binding_requirements)
+        return false unless state
+
+        state.possibilities.map(&:possibilities).flatten(1).any? do |poss|
+          activated.tag(:swap)
+          name = name_for(poss)
+          activated.set_payload(name, poss) if activated.vertex_named(name)
+          satisfied = binding_requirements.all? do |r|
+            requirement_satisfied_by?(r, activated, poss)
+          end
+          activated.rewind_to(:swap)
+          satisfied
+        end
+      end
+
+      # Filter's a state's possibilities to remove any that would not fix the
+      # conflict we've just rewound from
+      # @param [Hash] details of the conflict just unwound from
+      # @return [void]
+      def filter_possibilities_after_unwind(unwind_details)
+        return unless state && !state.possibilities.empty?
+
+        case unwind_details[:relationship]
+        when :primary then filter_possibilities_for_primary_unwind(unwind_details)
+        when :parent then filter_possibilities_for_parent_unwind(unwind_details)
+        end
```

Comment:
please add an `else` case that either is an explicit `nil` or raises an exception saying "invalid relationship type"